wai-extra/Network/Wai/Parse.hs

```diff
@@ -172,76 +306,128 @@ parseContentType a = do
+        Just rbt -> sinkRequestBodyEx o s rbt (requestBody r)
 
 sinkRequestBody :: BackEnd y
                 -> RequestBodyType
                 -> IO S.ByteString
                 -> IO ([Param], [File y])
-sinkRequestBody s r body = do
-    ref <- newIORef (id, id)
+sinkRequestBody = sinkRequestBodyEx noLimitParseRequestBodyOptions
+
+sinkRequestBodyEx :: ParseRequestBodyOptions
+                  -> BackEnd y
+                  -> RequestBodyType
+                  -> IO S.ByteString
+                  -> IO ([Param], [File y])
+sinkRequestBodyEx o s r body = do
+    ref <- newIORef ([], [])
     let add x = atomicModifyIORef ref $ \(y, z) ->
             case x of
-                Left y' -> ((y . (y':), z), ())
-                Right z' -> ((y, z . (z':)), ())
-    conduitRequestBody s r body add
-    (x, y) <- readIORef ref
-    return (x [], y [])
-
-conduitRequestBody :: BackEnd y
-                   -> RequestBodyType
-                   -> IO S.ByteString
-                   -> (Either Param (File y) -> IO ())
-                   -> IO ()
-conduitRequestBody _ UrlEncoded rbody add = do
+                Left y'  -> ((y':y, z), ())
+                Right z' -> ((y, z':z), ())
+    conduitRequestBodyEx o s r body add
+    (\(a, b) -> (reverse a, reverse b)) <$> readIORef ref
+
+conduitRequestBodyEx :: ParseRequestBodyOptions
+                     -> BackEnd y
+                     -> RequestBodyType
+                     -> IO S.ByteString
+                     -> (Either Param (File y) -> IO ())
+                     -> IO ()
+conduitRequestBodyEx o _ UrlEncoded rbody add = do
     -- NOTE: in general, url-encoded data will be in a single chunk.
     -- Therefore, I'm optimizing for the usual case by sticking with
     -- strict byte strings here.
-    let loop front = do
+    let loop size front = do
             bs <- rbody
             if S.null bs
                 then return $ S.concat $ front []
-                else loop $ front . (bs:)
-    bs <- loop id
+                else do
+                    let newsize = size + S.length bs
+                    case prboMaxParmsSize o of
+                        Just maxSize -> when (newsize > maxSize) $
+                            error "Maximum size of parameters exceeded"
+                        Nothing -> return ()
+                    loop newsize $ front . (bs:)
+    bs <- loop 0 id
     mapM_ (add . Left) $ H.parseSimpleQuery bs
-conduitRequestBody backend (Multipart bound) rbody add =
-    parsePieces backend (S8.pack "--" `S.append` bound) rbody add
+conduitRequestBodyEx o backend (Multipart bound) rbody add =
+    parsePiecesEx o backend (S8.pack "--" `S.append` bound) rbody add
+
 
-takeLine :: Source -> IO (Maybe S.ByteString)
-takeLine src =
-    go id
+-- | Take one header or subheader line.
+-- It makes sense to limit the maximum line length.
+-- Apache's default is 8190 (http://httpd.apache.org/docs/2.2/mod/core.html#limitrequestline)
```

Comment:
Perhaps this comment should be moved above to the actual usage of `8190` in the code?